### PR TITLE
[Web] Barcode Scanner

### PR DESCRIPTION
**Pull request by EvanBacon**
# How

Added a web worker which runs QR.js in the `expo-camera` module. Works when bundled and deployed [preview](https://quirky-wing-505041.netlify.com/) (point a QR at it and the QR will be outlined in blue)

# Test Plan

Works with native-component-list's Camera screen.

### CODE REVIEW

File: packages/expo-camera/package.json [hunk 1]
```diff
@@ -40,8 +40,10 @@
     "unimodules-permissions-interface": "*"
   },
   "dependencies": {
+    "jsqr": "^1.2.0",
     "lodash.mapvalues": "^4.6.0",
-    "prop-types": "^15.6.0"
+    "prop-types": "^15.6.0",
+    "qr-scanner": "^1.1.1"
```
Comment (bbarthec) [inline packages/expo-camera/package.json]: This one for sure is unused anywhere.

File: packages/expo-camera/package.json [hunk 2]
```diff
@@ -40,8 +40,10 @@
     "unimodules-permissions-interface": "*"
   },
   "dependencies": {
+    "jsqr": "^1.2.0",
```
Comment (bbarthec) [inline packages/expo-camera/package.json]: From what I see you load this library from some server.

File: packages/expo-camera/src/CameraModule/barcode/BarcodeScannerWorker.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+export default () => {
+  // @ts-ignore
+  self.importScripts('https://unpkg.com/jsqr@1.2.0/dist/jsQR.js');
```
Comment (bbarthec) [inline packages/expo-camera/src/CameraModule/barcode/BarcodeScannerWorker.ts]: Shouldn't this library be loaded from local `node_modules`?

**Review by ide — CHANGES_REQUESTED** [review]
I think the worker interface needs to be different -- toStringing the function and re-evaluating it in the worker context is too fragile (different set of globals, imports won't work). The webpack build needs to be configured to create a separate worker bundle.

The worker bundle can potentially be preloaded by the HTML page and by the service worker, but it should be a separate bundle since it runs in a different environment than the main page's JS.

File: packages/expo-camera/src/CameraModule/barcode/BarCodeScannerWorker.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/* eslint-env worker */
+
+declare var jsQR: Function;
+
+export default () => {
+  // @ts-ignore
+  self.importScripts('https://unpkg.com/jsqr@1.2.0/dist/jsQR.js');
```
Comment (bbarthec) [inline packages/expo-camera/src/CameraModule/barcode/BarCodeScannerWorker.ts]: What's our approach towards importing libs from outside like here? Do we want sth like this or rather we would like to ship them ourselves? 🤔

File: packages/expo-camera/src/ExponentCamera.web.tsx [hunk 5]
```diff
@@ -39,6 +40,49 @@ export default class ExponentCamera extends React.Component<NativeProps> {
...
+  };
+
+  private updateScanner = () => {
+    if (!this.camera) return;
+    const { barCodeScannerSettings, onBarCodeScanned } = this.props;
+    if (onBarCodeScanned && barCodeScannerSettings) {
+      this.camera.barCodeScanner.startScanner(
+        {
+          // Default barcode scanning update interval, same as is defined in the API layer.
+          // TODO: Bacon: Make this larger for low-end devices.
```
Comment (bbarthec) [inline packages/expo-camera/src/ExponentCamera.web.tsx]: ping

File: packages/expo-camera/src/CameraModule/barcode/BarCodeScannerWorker.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      const { image, types, options } = payload;
+      if (!image || !Array.isArray(types)) {
+        // @ts-ignore
+        postMessage(results);
+        return;
+      }
+
+      for (const type of types) {
+        switch (type) {
+          // TODO: Bacon: Add pdf417
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-camera/src/CameraModule/barcode/BarCodeScannerWorker.ts]: Probably best to support more formats in a new PR

File: packages/expo-camera/src/CameraModule/barcode/BarCodeScannerWorker.ts (hunk 4 above)
Comment (EvanBacon) [reply packages/expo-camera/src/CameraModule/barcode/BarCodeScannerWorker.ts]: Not sure, I'll need to do more research to figure out how to make this work in a concise manner.

**Comment by JesseRWeigel** [comment]
What is the status on this? Is there anything I could help with? I would like to use it for an upcoming project if possible.

**Comment by bbarthec** [comment]
@JesseRWeigel, the most suitable person to answer your question would be @EvanBacon 😉

**Comment by Nevtep** [comment]
Echoing JesseRWeigel, any chance this gets merged soon? how can I help?

**Comment by tsapeta** [comment]
@EvanBacon could you take care of this? It's time to merge/close such old pull requests 😄

**Comment by EvanBacon** [comment]
We still haven't decided how we want to handle workers in packages.

**Comment by dev-reactjs** [comment]
Hi @EvanBacon, i need to implement QR code scanner functionality in react-native and web through expo, current i'm using official expo-barcode-scanner which works in mobile fine but it breaks my web version ( expo start:web ), I tried to find alternative but not getting any success yet. Please suggest me when above thing merge or please let me know an alternative way. Because i can't found any relevant stuff. Please help.

**Comment by LucasLpr** [comment]
Any alternative ?

**Comment by ajimenezdev** [comment]
I'm using `react-qr-reader` when rendered on web. It is working with no problems

File: packages/expo-camera/src/CameraModule/barcode/BarCodeScanner.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  private scanForBarcodes(types: string[], config: CameraPictureOptions = {}) {
+    if (!this.webWorker) {
+      throw new Error('Cannot process a barcode before the worker has been created.');
+    }
+    const image = Utils.captureImageData(this.videoElement, config);
+    this.invokeWorker({ image, types, options: { inversionAttempts: 'dontInvert' } });
+  }
+
+  private invokeWorker(payload: any) {
+    if (!this.webWorker || !payload) return;
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/barcode/BarCodeScanner.ts:24]:
```suggestion
    if (!this.webWorker || !payload) {
      return;
    }
```

**Review by lukmccall — APPROVED** [review]

**Comment by EvanBacon** [comment]
related https://github.com/expo/expo/pull/9558

**Comment by alexkyse1** [comment]
When it will be released?

**Comment by captainR0bbo** [comment]
Is there any update on this?

**Comment by oussamabouchareb** [comment]
any plans to release this soon?

**Comment by tsapeta** [comment]
@oussamabouchareb
It's been merged on August, since that time we published a few new versions so I believe these changes are already included in the newest version 😉 Let me know if I'm wrong or something is not working as expected! 😄

**Comment by oussamabouchareb** [comment]
I Believe on expo's documentation it still says the web implementation is pending

https://docs.expo.io/versions/latest/sdk/bar-code-scanner/

But the camera component which I later switched to is also able to work on the web and scan barcodes just fine.

I apologise for wasting your time.

**Comment by mmalkush** [comment]
I got a message "BarCodeScanner Component not supported on the web" instead of scanner
"expo-barcode-scanner": "10.1.0"

@oussamabouchareb can you, please, share more details on how you make it work on the web?

**Comment by oussamabouchareb** [comment]
@mmalkush instead of using the BarCodeScanner component use the [Camera Componenet](https://docs.expo.io/versions/v40.0.0/sdk/camera/) it works on everything except simulators and it can also scan barcodes and QRcodes

**Comment by JJ810** [comment]
@oussamabouchareb, does it work on the web as well? I am using expo-camera but barcode scanning doesn't work on web.
It's working fine on iOS and Android though.

**Comment by EvanErickson** [comment]
@oussamabouchareb Can you share your code for the expo-camera that can scan barcodes? How did you add the function ,etc?

**Pull request merged**